### PR TITLE
AMDGPU/GlobalISel: Handle inreg arguments as SGPRs

### PR DESCRIPTION
**Pull request by arsenm**
This is the missing GISel part of 54470176afe20b16e6b026ab989591d1d19ad2b7

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-function-args.ll [hunk 1]
```diff
@@ -2777,4 +2777,462 @@ define void @vector_ptr_in_struct_arg({ <2 x ptr addrspace(1)>, <2 x ptr addrspa
...
+  ; CHECK-NEXT:   [[BUILD_VECTOR:%[0-9]+]]:_(<2 x p3>) = G_BUILD_VECTOR [[COPY]](p3), [[COPY1]](p3)
+  ; CHECK-NEXT:   [[DEF:%[0-9]+]]:_(p1) = G_IMPLICIT_DEF
+  ; CHECK-NEXT:   G_STORE [[BUILD_VECTOR]](<2 x p3>), [[DEF]](p1) :: (store (<2 x p3>) into `ptr addrspace(1) undef`, addrspace 1)
+  ; CHECK-NEXT:   SI_RETURN
+  store <2 x ptr addrspace(3)> %arg0, ptr addrspace(1) undef
+  ret void
+}
+
 attributes #0 = { nounwind }
+
```
Comment (rampitec) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-function-args.ll:3237]: Drop the empty lines.

**Review by rampitec — APPROVED** [review]
LGTM with a nit.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-function-args.ll [hunk 2]
```diff
@@ -2777,4 +2777,462 @@ define void @vector_ptr_in_struct_arg({ <2 x ptr addrspace(1)>, <2 x ptr addrspa
...
+  store i1 %arg0, ptr addrspace(1) undef
+  ret void
+}
+
+define void @void_func_i8_inreg(i8 inreg %arg0) #0 {
+  ; CHECK-LABEL: name: void_func_i8_inreg
+  ; CHECK: bb.1 (%ir-block.0):
+  ; CHECK-NEXT:   liveins: $sgpr0
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:_(s32) = COPY $sgpr0
```
Comment (jwanggit86) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-function-args.ll:2799]: For some GPUs, isn't it true that sgpr0-spgr3 are reserved?

File: llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-function-args.ll
Comment (jwanggit86) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-function-args.ll:1]: Maybe add a testcase where the input args exhaust the available SGPRs?

**Review by jwanggit86 — COMMENTED** [review]
(1) Are sgpr0-sgpr3 reserved for some GPUs?
(2) A testcase where input args exhaust available SGPRs.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-function-args.ll (hunk 2 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-function-args.ll:2799]: Yes, this is starting at the wrong location

File: llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-function-args.ll
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-function-args.ll:1]: That breaks, see issue #78121

**Comment by arsenm** [comment]
The implicit argument handling is pretty broken, so will fix that separately

**Pull request merged**